src/main/java/com/puppycrawl/tools/checkstyle/xpath/AbstractNode.java

```diff
@@ -0,0 +1,361 @@
+import net.sf.saxon.type.SchemaType;
+
+/**
+ * Represents general class for {@code ElementNode}, {@code RootNode} and {@code AttributeNode}.
+ *
+ * @author Timur Tibeyev
+ */
+public abstract class AbstractNode implements NodeInfo {
+
+    /** The children. */
+    private final List<AbstractNode> children = new ArrayList<>();
+
+    /**
+     * Getter method for token type.
+     * @return token type
+     */
+    public abstract int getTokenType();
+
+    /**
+     * Returns underlying node.
+     * @return underlying node
+     */
+    public abstract DetailAST getUnderlyingNode();
+
+    /**
+     * Getter method for children.
+     * @return children list
+     */
+    protected List<AbstractNode> getChildren() {
+        return Collections.unmodifiableList(children);
+    }
+
+    /**
+     * Add new child node to children list.
+     * @param node child node
+     */
+    protected void addChild(AbstractNode node) {
+        children.add(node);
+    }
+
+    /**
+     * Returns true if nodes are same, false otherwise.
+     * @param nodeInfo other node
+     * @return {@code TreeInfo}
+     */
+    @Override
+    public boolean isSameNodeInfo(NodeInfo nodeInfo) {
+        return this == nodeInfo;
+    }
+
+    /**
+     * Returns if implementation provides fingerprints.
+     * @return {@code boolean}
+     */
+    @Override
+    public boolean hasFingerprint() {
+        return false;
+    }
+
+    /**
+     * Returns uri of the namespace for the current node.
+     * @return uri
+     */
+    @Override
+    public String getURI() {
+        return "";
+    }
+
+    /**
+     * Returns if current node has children.
+     * @return if current node has children
+     */
+    @Override
+    public boolean hasChildNodes() {
+        return !children.isEmpty();
+    }
+
+    /**
+     * Determines axis iteration algorithm.
+     * @param axisNumber element from {@code AxisInfo}
+     * @param nodeTest filter for iterator
+     * @return {@code AxisIterator} object
+     */
+    @Override
+    public AxisIterator iterateAxis(byte axisNumber, NodeTest nodeTest) {
+        AxisIterator axisIterator = iterateAxis(axisNumber);
+        if (nodeTest != null) {
+            axisIterator = new Navigator.AxisFilter(axisIterator, nodeTest);
+        }
+        return axisIterator;
+    }
+
+    /**
+     * Compares current object with specified for order.
+     * @param nodeInfo another {@code NodeInfo} object
+     * @return number representing order of current object to specified one
+     */
+    @Override
+    public int compareOrder(NodeInfo nodeInfo) {
+        return getLocalPart().compareTo(nodeInfo.getLocalPart());
+    }
+
+    /**
+     * Returns namespace array.
+     * @param namespaceBindings namespace array
+     * @return namespace array
+     */
+    @Override
+    public final NamespaceBinding[] getDeclaredNamespaces(NamespaceBinding... namespaceBindings) {
+        throw throwUnsupportedOperationException();
```

Comment:
all throwUnsupportedOperationException in AbstractNode should be explained , to make it clear why we do not need that methods at all. Why non of child classes need it.